### ISSUE TITLE
Better map event distribution

### ISSUE THREAD

**Issue by Turtyo** [Saplings-Projects/1M_sub#108]
## Is your amelioration related to a problem? Please describe.

The distribution of events is truly random currently (1/5 to get a certain event on any given room), which could end up with 4 heal / shop event in succession. This is clearly not what we want. it worked until now for testing, but we need to get closer to what it should actually feel like

## Describe the solution you'd like
We need a way to have a better control on what event spawn to what event. For example it shouldn't be possible to have a succession of shop/heal/shop or shop/shop or heal/heal for example.
In a way, what's in the room around has to influence the possibilities of what you get in the current room. I'm sure there exist plenty of algorithms to do exactly this kind of thing, but I have not researched any yet.

## Describe alternatives you've considered
A more simple alternative is simply have weighted probabilities on what rooms can appear, like 60% mob, 20% dialogue, 10% shop, 5% heal and 5% random.
But this causes more problem, for example it could end up with not having any shop or heal on the entire floor due to bad luck, which is probably something we want to avoid. We want something that looks random but has a set of rule to it (like for example at least 3 shops on the floor put in certain positions to ensure the player will always find at least one) 

## Additional context
The kind of maps we can have currently, only a single mob event in the first three floors, but two heals and 3 shops
![image](https://github.com/Saplings-Projects/1M_sub/assets/49365680/ab6b5e4c-f43b-4e8f-b6e9-7f7fe38e3427)

**Comment by TripleCubes** [comment]
Get all room positions in the first 3 floors, shuffle it, the first 3 in array is heal, next 3 is shop, next 3 is ... you get the idea

And then, get the minimum number of room player need to go through to get through 3 floors (which is, 3), one of them must be heal

So like, for example, room number 2 that player go through will be guaranteed to be heal. It can be room 1 or 3 too, choosen randomly

The way we choose it is simple, get a list of steps [1, 2, 3], shuffle the list, the first index will be the heal room. If we want 2 heal rooms get the first two indexes

There is a problem tho, a forced event (like shop) can override a heal room. The solution for this is, or we can just leave it like that and add more heal rooms, or if a forced event overlap with a heal room, we place that heal room on a non-important room, like a battle room

**Comment by Turtyo** [comment]
Shuffling like that might also result in unwanted distributions if you are unlucky, it doesn't change much from randomly selecting one for each room.
As I said, it will be better to check if an algorithm for distributions such as this already exists (which I'm pretty sure it does, considering how often this kind of stuff needs to be done in games).
I'll check that more in depth tonight when I have the time

**Comment by TripleCubes** [comment]
How is shuffling like this cause unwanted distribution? The minimum number of room is guaranteed, and the player is guarateed to always get a number of certain rooms

Isnt that what we want?

**Comment by Turtyo** [comment]
Let's say we choose to have 3 heal events. You shuffle, and the 3 heal event end next to each other.
How do you prevent that from happening ? Just because it's unlikely to happen doesn't mean it can't happen. I took the example with 3 heal, but since we have heal + shop there's actually more of those events which we don't want next to each other.
Since the map is dark, we might also want to not put every event of the same type close to each other (even if not immediately close).

For example if the player chooses to go on the right of the map but all the heal events ended up on the left, then the player doesn't get a single heal because of bad luck. You can't control how the events are distributed with just a shuffle.

Even if we take the addition you make to the shuffle, which is to force a spawn, it's not good imo because on the opposite, it makes it feel way less random. And we might actually not want to spawn any heal event in the first three floor.

**Comment by Turtyo** [comment]
I found [this blog post on Steam](https://steamcommunity.com/sharedfiles/filedetails/?id=2830078257) about map generation in StS. We can ignore the path generation since the shape of our map is determined beforehand.
I think starting from something close to that will make it so that we are on a good basis

**Comment by Lann-hyl** [comment]
Since the map is tree shaped, is it possible for nodes/rooms to have parent-child relationship? Like we create the child room and generate its type with rules depending on its parent types
Also do we need to change the map generation algorithm in the future or are we good? The one in the blog is pretty good but maybe unnecessary

**Comment by Turtyo** [comment]
I don't think we need parent-child relations, since we can already know that based on the position of room
if x is the position on the floor and y the number of the floor knowing that x doesn't count the index of the room, but the actual position relative to the side of the map), then from a room of position (x,y), we can only go to rooms of position (x-1,y+1), (x, y+1), (x+1, y+1)
Little drawing:
![image](https://github.com/Saplings-Projects/1M_sub/assets/49365680/70fa13ca-0859-47db-89a7-0e7e00cc627d)

We could make the check deeper, but it's probably a big complexity. The map is not big, but it might be needed to generate it in the background just to be sure. I think their algorithm to put events already prevents that type of shop-shop scenario. They accept shop-mob-shop scenario it seems like, I am unsure why. Maybe it doesn't really matter, but that's probably dependent on the game. 

If you look at how Inscryption works for example, you carry fur until the end of the level where you can trade it with the fur trader, and there is only a single one per level, but the things you can get are really powerful. In StS you can get powerful things, but a single item of StS is generally less impactful than a single trader bonus of Inscryption.

Which map generation algorithm are you talking about ? The one about creating the links ? Or the one about putting the events ? We don't need to one for creating the links, as the shape of our map is already decided. We might better the algorithm to generate the rooms in the future to make it a bit more random (since right now the shape is chosen, but we might want to generate it, the rules to do so are pretty simple).

**Comment by Lann-hyl** [comment]
Yeah I saw that the map layout is like on the 1st image, each child room has at most 3 parents, so it is quite easy to scan.
![Screenshot 2024-06-08 201948](https://github.com/Saplings-Projects/1M_sub/assets/43467955/8a786faa-e8d3-473d-9479-a388e8da891f)

I was talking about the one creating the rooms and the links/paths, sorry for not being clear. I saw that we fix the width, then pad ->add rooms -> pad. I was wondering if we wanted later to have a more complex layout like on this image for example:
![image](https://github.com/Saplings-Projects/1M_sub/assets/43467955/35772a9a-d8a3-4189-9f4b-f4762b36081c)
where we'd have spaces with no rooms, and some rooms with no paths in between (like the one from (4,2) to (3,3)) just for some diversity and complexity.

Spaces with no rooms is just a null check, but for the second case, we wouldn't be able to just scan.
But I guess that one easy solution is simply that if a room is in (x-1,y+1), (x, y+1), (x+1, y+1) position of our current room, then there is a path in between and the second case shouldn't exist.

**Comment by Lann-hyl** [comment]
@Turtyo Hey has this issue been approved and can I assign myself to this issue? Since it is more about algorithms and less Godot specific it kinda works well for me. We can discuss more about the methods we want to implement (like the probabilities, etc...) when you have time.

**Comment by Turtyo** [comment]
we don't want to do this
![image](https://github.com/Saplings-Projects/1M_sub/assets/49365680/70f2ba2d-a777-46b3-b2b5-7098b4de6b8a)

We want to keep it like this
![image](https://github.com/Saplings-Projects/1M_sub/assets/49365680/7fb9cf80-2b27-44c8-a927-2acdd3141a2f)

The only thing that can make a more complex layout is having randomized map width. This is what I meant by "We might better the algorithm to generate the rooms in the future to make it a bit more random". There are just a few rules to follow to make the randomization of the map width work correctly, but that's not the subject right now and it would be more appropriate in a new issue.

TL;DR : **we do not want not want to use their path generation algorithm**

To see if an issue has been approved, you can just check the tags of the issue. As you can see, this issue hasn't be approved yet
![image](https://github.com/Saplings-Projects/1M_sub/assets/49365680/59ed50ec-88ab-4ecd-98d8-90047a7b9947)

We can finish discussing the implementation we want before approving it, as having everyone on the same page is more important than starting faster.

Regarding probabilities, we can start with something similar to what they have, since we don't have Elite (actually we have a mini-boss but we already discussed the position, somewhere around the middle of the map, we just haven't put it there), we can distribute the probability attributed to elites around the 4 other types (so 4% to each other event type)
- mob: 49%
- Dialogue: 26%
- Rest: 16%
- Merchant: 9%

Total is indeed 100%. Regarding the 4 rules used by StS:
![image](https://github.com/Saplings-Projects/1M_sub/assets/49365680/4dab02b3-0b78-4f4e-bd1f-e81968fec041)

 1. Since there are 14 floors, and heal only spawn starting floor 7, that's no heal before half of the map. So for us we use this criteria, no heal before the first half. We have less than 14 floors anyway
 2. No changes there
 3. That seems harder to uphold, because we have less event types and 3 path per room (instead of 2), so we might have to change this, idk know how though
 4. Basically no rest on the floor that's 2 floors before the boss

I'm not sure to understand the sentence "If one of the rules is broken it will reassign a Location based on the previously mentioned odds until it obeys by all the above rules."
By reassign a location, does this mean it swaps things ? Or does it do a new pick of event on the rooms that don't obey the rules ? Solution one could fail (because there could be no solution). Solution two might shift the actual probabilities.

**Comment by Lann-hyl** [comment]
Ok got it for the path generation, I was just wondering for future plans if there was any.

For the rules:
1. Yes we could adapt it to any floor number anyway, and what difficulty we seek
2. -
3. I don't think this rule is really necessary either, I don't think slay the spire respects it either(?). If we really want to uphold it each room will have to check its neighbours
4. I wonder if thats too harsh but that'll depend on difficulty

"If one of the rules is broken it will reassign a Location based on the previously mentioned odds until it obeys by all the above rules." -> I think this means when a room is created, the room type is redrawn until all rules are respected (so solution 2). 
Swapping seems a weird thing to do because we could just never reach a solution that satisfies all rules.

I think its ok if the generated map does not exactly have the same distribution as our probabilities (so not exactly 50% of room being monsters, etc...) as long as we don't get a scenario where we only have 1 shop/rest spot in the whole map.

**Comment by Tomzkk** [comment]
> 3. A room that has 2 or more Paths going out must have all destinations be unique. 2 destinations originating from the same Room cannot share the same Location
> That seems harder to uphold, because we have less event types and 3 path per room (instead of 2), so we might have to change this, idk know how though

I am not sure if I am understanding this rule correctly, and it's been a while since I had StS open, but I went to look at a few maps and I am pretty sure StS does not uphold this rule at all ? 
![image](https://github.com/Saplings-Projects/1M_sub/assets/24390212/ecc61eb9-47f9-4a3c-9480-e6c72fae5dc1)
![image](https://github.com/Saplings-Projects/1M_sub/assets/24390212/1fea5e2e-da75-45ee-9d92-b5b21cb172cc)


> "If one of the rules is broken it will reassign a Location based on the previously mentioned odds until it obeys by all the above rules."

There is an example image on the page which shows what is supposed to happen according to them. It seems the rooms that were an issue get reassigned to different types, the rest stays the same.
![image](https://github.com/Saplings-Projects/1M_sub/assets/24390212/accbe2f3-8989-4850-a3bb-3ef05651398c)

**Comment by Turtyo** [comment]
> but for implementation the type reassignment should be when the room is created, we don't need to create the whole map.

The entire map is created at the start though. For the simple reason that you never know when the player will put down a torch, so it's better to have the entire map ready (and it's not that big anyway)

I was also thinking that we have the random event. So the % i gave for events are wrong, we should replace the elite by the random event, which gives:
- mob: 45%
- Dialogue: 22%
- Random: 16% (you don't see the event until you enter the room)
 - Rest: 12%
 - Merchant: 5%

And the random event you have 1/4 to get any of the other events currently.

The map generation is done in the `MapManager`. Creating the shape of the map and putting the events is done in the same function. I think it would be good to have two functions, one that creates the shape of the map, and another that puts the events in. First function can just put `null` inside the `room_event` and the second function is the one taking care of filling the events.

This will probably be cleaner like that, and later on we can modify the algorithm those choose the generation of the map width without any influence of the event filling.

By parent do you mean the floor above the current floor ? I also think it would be good to not have 3 times the same event available from a given room. I think generating the whole map then checking will work

**Comment by Lann-hyl** [comment]
> it's better to have the entire map ready
I realize now I worded it poorly, but yes we are on the same page

> I think it would be good to have two functions
Yes even for readability it's better

By parent I meant the rooms of the previous floor. I've been using the terminology like this but idk if we have other terms:
![image](https://github.com/Saplings-Projects/1M_sub/assets/43467955/57cff50a-3b87-4ec0-a51a-80fcf49bc851)

**Comment by Tysterman74** [comment]
Off-topic on what y'all been cooking up in this discussion, but I want to also add a debug flag to turn off the lighting mechanic so we can fully see the map for testing. I don't know if we want a separate issue on this, but I can add it in real quick if that's what we want.

**Comment by Lann-hyl** [comment]
Yes that'd be helpful to debug, not sure either if we have to make a separate issue.

Also a suggestion for rule 3 if it is too strict, we can relax it like this:
Rule 3: A Room that has 2 or more Paths going out must have all destinations be unique.
-> Rule 3 Relaxed: There must be at least 2 room types among destinations of Rooms that have 2 or more Paths going out.
The only difference is that the scenario in this image is also possible now, less strict but we keep some diversity. The rest is unchanged.
![image](https://github.com/Saplings-Projects/1M_sub/assets/43467955/f743d198-aeb6-42a3-9481-dabe628d871e)

But that's just food for thought, we can implement what we have decided first then see how it works out.

**Comment by Turtyo** [comment]
@Lann-hyl I agree with the change for rule three. Now for the technical aspect, it's mainly following the algorithm and modifying the MapManager, I think you will manage. You can start if you want

@Tysterman74 that would be helpful indeed. Maybe Lann can do it at the same ? Were the dark areas just a plan over the map, and if so can we just disable it directly ? If it's easy like that, I don't think there is a need for a new issue, as it will be helpful right now and it's better to have right away

**Comment by Lann-hyl** [comment]
For some reason, the _ready() function of MapManager is called whenever we enter a monster room. I don't know if it's normal, I guess its Godot switching scenes or something(?). Just in case I'll put a `if !is_map_initialized()` at the start so that it doesn't recreate a map everytime.

**Comment by Turtyo** [comment]
That's a very strange behavior, I never noticed that; I'll check on my side to see if I can recreate that.

**Comment by Turtyo** [comment]
Indeed I have the same thing happening. This is strange, as the MapManager is a singleton, meaning it's loaded globally and isn't reloaded between events. This means that something somewhere is calling the `_ready` of the MapManager in some sort (I checked the SceneManager for example, its ready function is only called once)

You will have to see why the `_ready` is called each time, it seems to come from the scene switching but at the same time, there are sometimes strange things that happen due to signals being emitted so it's not always easy to find the reason
